doctor/doctor.go

```diff
@@ -0,0 +1,240 @@
+)
+
+var log = logging.Logger("namesys")
+
+type Doctor struct {
+	host    libp2p.Host
+	config  *config.Config
+	console io.Writer
+
+	mu                     sync.RWMutex
+	sub                    libp2pEvent.Subscription
+	reachability           libp2pNetwork.Reachability
+	hasNagged, usingRelays bool
+}
+
+type Status struct {
+	Reachability       libp2pNetwork.Reachability
+	AutoRelayEnabled   bool
+	UsingRelays        bool
+	Listening          bool
+	TCPPorts, UDPPorts []int
+	LocalIP            net.IP
+	Gateway            *url.URL
+}
+
+func NewDoctor(host libp2p.Host, cfg *config.Config, console io.Writer) *Doctor {
+	return &Doctor{
+		host:    host,
+		config:  cfg,
+		console: console,
+	}
+}
+
+func (n *Doctor) Close() error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.sub == nil {
+		return nil
+	}
+	_ = n.sub.Close()
+	n.sub = nil
+	return nil
+}
+
+func (n *Doctor) GetNATStatus() libp2pNetwork.Reachability {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	return n.reachability
+}
+
+func (n *Doctor) GetRelayStatus() bool {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	return n.usingRelays
+}
+
+func (n *Doctor) Start() error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.sub != nil {
+		return fmt.Errorf("already started")
+	}
+
+	var err error
+	n.sub, err = n.host.EventBus().Subscribe([]interface{}{
+		new(libp2pEvent.EvtLocalReachabilityChanged),
+		new(libp2pEvent.EvtLocalAddressesUpdated),
+	})
+	if err != nil {
+		return err
+	}
+
+	evtCh := n.sub.Out()
+	go func() {
+		defer n.Close()
+
+		for evt := range evtCh {
+			switch evt := evt.(type) {
+			case libp2pEvent.EvtLocalReachabilityChanged:
+				n.handleReachability(evt)
+			case libp2pEvent.EvtLocalAddressesUpdated:
+				n.handleAddrUpdate(evt)
+			default:
+				log.Errorf("unexpected event type: %T", evt)
+			}
+		}
+	}()
+	return nil
+}
+
+func (n *Doctor) printf(f string, rest ...interface{}) {
+	fmt.Fprintf(n.console, f+"\n", rest...)
+}
+
+func (n *Doctor) handleAddrUpdate(evt libp2pEvent.EvtLocalAddressesUpdated) {
+	usingRelays := false
+	for _, addrUpdate := range evt.Current {
+		if _, err := addrUpdate.Address.ValueForProtocol(ma.P_CIRCUIT); err == nil {
+			usingRelays = true
+			break
+		}
+	}
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if usingRelays != n.usingRelays {
+		n.usingRelays = usingRelays
+		status := "not using relays"
+		if usingRelays {
+			status = "using relays"
+		}
+		n.printf("Relay Status: %s", status)
+	}
+}
+
+func (n *Doctor) handleReachability(evt libp2pEvent.EvtLocalReachabilityChanged) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if evt.Reachability == n.reachability {
+		// Otherwise, we'll print
+		// "unknown" on start.
+		return
+	}
+	n.reachability = evt.Reachability
+	status := "unknown"
+	switch n.reachability {
+	case libp2pNetwork.ReachabilityPrivate:
+		status = "private"
+	case libp2pNetwork.ReachabilityPublic:
+		status = "public"
+	default:
+	}
+
+	n.printf("Firewall Status: %s", status)
+	if !n.hasNagged {
```

Comment:
@stebalien Would you remember why you put in this flag ? What's the use case ?

Reply:
it appears to be so that it only says it once, and then remembers that it has said it so it doesn't say it again, right?